### PR TITLE
Reconcile resources with generations

### PR DESCRIPTION
**Pull request by ashleyschuett**
**What this PR does / why we need it**:

Reconcile will update the resource to our desired resource if labels or annotations do not match, or if the generation Id is different from the generation Id that has been saved as the last generation on the KubeVirt CR. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partial Fix # https://bugzilla.redhat.com/show_bug.cgi?id=1868099

**Special notes for your reviewer**:

**Release note**:
```release-note
Reconcile Deployments, DaemonSets, MutatingWebhookConfigurations and ValidatingWebhookConfigurations
```

### CODE REVIEW

**Comment by ashleyschuett** [comment]
/hold 

going to split this out into multiple PRs the first of which is https://github.com/kubevirt/kubevirt/pull/4788

**Comment by ashleyschuett** [comment]
@rmohr this PR is finally ready for review. It uses the generation field to determine if a reconciliation is needed.

File: manifests/generated/operator-csv.yaml.in
Comment (stu-gott) [inline manifests/generated/operator-csv.yaml.in:225]: Be careful with this. Are we sure we don't need patch? E.g. impact on upgrade??

File: pkg/virt-operator/resource/apply/admissionregistration.go
Comment (stu-gott) [inline pkg/virt-operator/resource/apply/admissionregistration.go:90]: So in general, the advantage to `Patch` is that it's atomic. Sanity checks and update are performed in one step. `Update` by nature is prone to race conditions.

Also, it's usually a very good idea to use `Patch` when you don't "own" the type. In other words, `ValidatingWebhookConfigurations` was defined in somebody else's codebase. While it is true that Kubernetes is very careful about updating their API--it can happen. Where this can go very poorly is the way JSON marshalling is handled. Let's pretend that a field was added to the `ValidatingWebhookConfigurations` object (which can happen if Kubernetes is updated and KubeVirt isn't). If our client gets a copy of this object, it'll throw out this new field. When we update later, there's an inadvertent side effect--we effectively delete whatever value that field had in it.

So with that in mind, is there any way possible to use `Patch` here instead of `Update`?

**Review by stu-gott — COMMENTED** [review]
Looks good overall. Just some concerns about Update vs Patch

File: pkg/virt-operator/resource/apply/admissionregistration.go
Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/admissionregistration.go:90]: Switching this back to using `Patch` is defiantly possible if that is preferred. I'm not sure I understand the case you described however. Since we are using the `Patch` operation "replace" for the full `spec` / `workloads` key wouldn't Patch and Update react the same? Unless you are referring to a top level key being added/changed? 

`os = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec", "value": %s }`, string(newSpec)))`

**Review by stu-gott — COMMENTED** [review]
/lgtm

File: pkg/virt-operator/resource/apply/admissionregistration.go
Comment (stu-gott) [reply pkg/virt-operator/resource/apply/admissionregistration.go]: Spoke offline with @ashleyschuett to clarify. This comment is for the benefit of other reviewers. The trouble with Update is it replaces the entire record, while Patch only updates specific fields of interest.

When using "Update", if in the future, a new field is added to an object being reconciled, then our code would silently ignore it (read: delete the value). When performing an update on the entire record this would effectively delete that field--which isn't our intention. The danger of this would be potentially hot-looping.

File: tests/operator_test.go
Comment (davidvossel) [inline tests/operator_test.go:878]: one thing I'm a little paranoid about is the operator getting into an update storm of some sort, where it's fast looping writing objects because it thinks they are out of date for some reason.

What would you think about writing some func test logic that verifies that the new generation list doesn't change after an update is complete. Maybe this could be added to the `waitForKvWithTimeout` function, something that just observes that no changes occur to the generation list for something like 15 seconds after the kv object looks like it's stable.

**Review by davidvossel — COMMENTED** [review]
I did one pass and the general idea seems solid. I need to give this a more detailed look though, which i'll try to get to this week. I had one thought about testing that I left a comment about.

File: tests/operator_test.go
Comment (ashleyschuett) [reply tests/operator_test.go]: Yeah, that makes sense. I've added a `Consistently` check at the end of this func test to check that the generation stays the same on the KV resource from the time the resource is updated back to it's original state.

File: pkg/virt-operator/kubevirt_test.go
Comment (davidvossel) [inline pkg/virt-operator/kubevirt_test.go:1719]: i see that you're adding to the kv.Status.Generations list in the normal add functions. Like `addDaemonSet()` for example.

I'm curious why you explicitly set the generations for these webhooks here when I'd think they'd already be added in the `addValidatingWebhook()` shared function.

File: pkg/virt-operator/kubevirt_test.go
Comment (davidvossel) [inline pkg/virt-operator/kubevirt_test.go:1683]: We have a test case here that verifies that nothing will change if a resource's generation does not change "should do nothing if KubeVirt object is deployed". That's great, we have that part covered.

I'd like to see a unit test that exercises the logic specifically around strictly reconciling a resource when the generation ID doesn't match what's on the kubevirt cr. 

That way we'll have covered both sides... the do nothing side (already covered now) and the should do something side

File: pkg/virt-operator/resource/apply/apps.go
Comment (davidvossel) [inline pkg/virt-operator/resource/apply/apps.go:78]: I've been pondering this for awhile. There's a (possibly benign) race between when we set this new generation on the kubevirt object and when we actually see the new generation on the kubevirt object in our cache.

So, for instance, if you did the following sequence, you'd get an interesting result.

pseudo code 

```
resourcemerge.SetDeploymentGeneration(&kv.Status.Generations, deployment)
client.Update(kv)
cachedKV := getKVFromCache(kvInformer, kv.Name, kv.Namespace)

if !reflect.DeepEquals(cachedKV, kv) {
     // cache doesn't reflect the latest generation changes.
}
```

I'm not sure what to think of this race condition though. Functionally, it might not cause any problems except that we may Patch the same object a few times with the same change until all the informers catch up. This is the same race that the "expectations" ref counting logic exists to solve.

If applying the same patch to an already patched object never results in the generation ID changing somehow, then I *think* the issue i'm describing here woudl just be an optimization that probably isn't worth our time. If however having the cache not synced with reality can result in generation id of a resource spinning, we'd be in trouble. 

I don't know if i did a good job explaining my thoughts here or not.  I'm about 80% sure it's not an issue. Have you thought about this at all?

**Review by davidvossel — COMMENTED** [review]
looks really solid. I'd just like to see one unit test that exercises the logic that occurs when generation IDs don't line up before it goes in.

**Review by davidvossel — COMMENTED** [review]
/approve
/lgtm

File: pkg/virt-operator/resource/apply/apps.go
Comment (davidvossel) [reply pkg/virt-operator/resource/apply/apps.go:78]: i don't think my concern here occurs in a way that causes an issue. worst case, we patch and already patched object. When that occurs, the test operation in the patch json will cause the operation to fail until our cache catches up.

Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/apps.go:78]: sorry, I was thinking through this more as well and then didn't submit my comment. 

I agree, the KV object is requested from cache once right before the reconcile loop and then updated at the end of it. Worst case should be that the object is patched, kubernetes sees that nothing is changed and no ops, then returns the same object with the same generation as it was. 

I'm not sure what you are referring to by the `test operation`, we don't add a test operation patch to the JSON patch for the four resources included in this PR. Is that something that should be added?

Comment (davidvossel) [reply pkg/virt-operator/resource/apply/apps.go:78]: ah! I think it would be wise to add a test op to the patch... Actually, the test operation can be to verify the generation of the cached object matches the generation in our kv.Status.Generations list. 

so something like

``` fmt.Sprintf(`{ "op": "test", "path": "/metadata/generation", "value": %s }`, expectedGenerationID)```

what do you think?

Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/apps.go:78]: wont that then fail when we actually want to update? User updates resource generation ID is increased but the expectedGeneration on the KV resource is not, we want to update the resource but the generation id in cache would then not match the expected and error? Should we instead add the expected generation of the cached object to the test patch?

**Review by davidvossel — COMMENTED** [review]
yup, i think adding the generation patch is a good safety mechanism here. thanks for doing that. 

i had one comment about the ordering of the test operation. Perhaps the test should come first in the operations list?

In the past we've put the test op first because it typically involves testing that a field is equal to what we have in cache before replace that field... Here it's a little different though, because we're testing that one field is equal to what we expect and then replacing another field. I'm not 100% sure if ordering matters in this case, but maybe it could somehow depending on how the patch is evaluated.

File: pkg/virt-operator/resource/apply/apps.go
Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/apps.go]: ahh yeah that makes sense. updated!

**Review by davidvossel — APPROVED** [review]
/lgtm

**Review by davidvossel — APPROVED** [review]
/lgtm

**Comment by ashleyschuett** [comment]
@davidvossel I have looked further into some of the flakeyness that this PR is failing on. While most of them are due to overall CI flakeyness. This PR will introduces flakeyness on the new function test that it adds. I think there are a few things that contribute to this. 

1. Having multiple threads for the kubevirt controller seems to be a problem. Since multiple objects can be being processed at the same time and then try to update the KubeVirt object status many will error and requeue due to trying to update an old kubevirt object at the end of the sync. I think having only one thread running would be better. 

2. There is now an issue with getting the KubeVirt object from the cache since I am getting the expected version of the deployments/ds/vwc/mwc from the status of the object, when it is old the expected generation will not equal the resource in cache causing a patch (which will increment the generation on the object) and then this process will go on several times until the cache finally returns the current state which can take a few passes. This goes away if I instead get the KubeVirt resource using the clientset instead of the informer. Which I think might be better? Calling the API for the KV resource vs multiple unneeded patch calls? 

Im curious what you think about this trade offs for performance if these solutions should be implemented or if there are other ways to mitigate these issues? Or maybe since Kubernetes should always be driving towards the desired state its okay for these multiple unneeded sync as that is the way the system is designed? and instead in the func test i should have a longer wait for the KubeVirt object generation to equal the expected generation and continuously get them both since the deployment could be patched and generation increased again after going back to the correct state..

**Comment by davidvossel** [comment]
> Having multiple threads for the kubevirt controller seems to be a problem. Since multiple objects can be being processed at the same time and then try to update the KubeVirt object status many will error and requeue due to trying to update an old kubevirt object at the end of the sync. I think having only one thread running would be better.

Even though we have multiple threads, the kubevirt object won't get processed multiple times in parallel. I think the issue you're hitting here is that the informers aren't catching up fast enough with all the changes. 

> There is now an issue with getting the KubeVirt object from the cache since I am getting the expected version of the deployments/ds/vwc/mwc from the status of the object, when it is old the expected generation will not equal the resource in cache causing a patch (which will increment the generation on the object) and then this process will go on several times until the cache finally returns the current state which can take a few passes. This goes away if I instead get the KubeVirt resource using the clientset instead of the informer. Which I think might be better? Calling the API for the KV resource vs multiple unneeded patch calls?

Yep, i get it. we need the informer cache to catch up or else we're going to be spinning trying to do things that ultimately fail. 

I think this can be optimized. here are some options. 

1. Cache the kubevirt CR's expected generations map locally and don't process the kubevirt CR again until the map in our cache matches the expected one. This is basically the same concept as our "expecter" logic. We're preventing the reconcile loop from executing until an expected change shows up in our cache. 

2. Throttle the kubevirt CR's reconcile loop. I actually do that [here](https://github.com/kubevirt/kubevirt/blob/9a19be79b130fca459a9cc730988d32b0ba0b781/pkg/virt-controller/watch/workload-updater/workload-updater.go#L375) in another controller. The issue is if lots and lots of objects can cause a reconcile loop to pop (like in virt-operator), then the reconcile loop get's thrashed with changes non stop. In the workload update reconcile loop, i ensure that the kubevirt CR only gets processed at most every 5 seconds. This avoids the controller executing computations non stop while 100s-1000s of vmis are in flight. 


While it's kind of lazy, I think the most pragmatic approach is to just throttle the reconcile loop so the KubeVirt CR only gets worked on once every few seconds.  That gives the informers time to catch up and prevents the loop from spinning like crazy as all the new object changes come in on the informers.

File: pkg/virt-operator/kubevirt.go
Comment (davidvossel) [inline pkg/virt-operator/kubevirt.go:87]: nice. I didn't realize the queue had this functionality. I need to update some of my code now in other places.

**Review by davidvossel — APPROVED** [review]
/lgtm

**Comment by ashleyschuett** [comment]
Hey @davidvossel can i get another `lgtm` on this PR, hopefully for the last time?

**Pull request merged**
